writeengine/bulk/we_bulkloadbuffer.cpp

```diff
@@ -1721,107 +1690,1381 @@ int BulkLoadBuffer::parseCol(ColumnInfo& columnInfo)
+            if (defLen > column.definedWidth)
+              memcpy(charTmpBuf, defData, column.definedWidth);
+            else
+              memcpy(charTmpBuf, defData, defLen);
+          }
+          else
+          {
+            idbassert(width <= 8);
+            for (int j = 0; j < width - 1; j++)
+            {
+              charTmpBuf[j] = '\377';
+            }
+            charTmpBuf[width - 1] = '\376';
+            pVal = charTmpBuf;
+            memcpy(p, pVal, width);
+            continue;
+          }
+        }
+        else
+        {
+          const uint8_t* data = binaryArray->GetValue(i, &tokenLen);
+          const char* dataPtr = reinterpret_cast<const char*>(data);
+          if (tokenLen > column.definedWidth)
+          {
+            uint8_t truncate_point = utf8::utf8_truncate_point(dataPtr, column.definedWidth);
+            memcpy(charTmpBuf, dataPtr, column.definedWidth - truncate_point);
+            bufStats.satCount++;
+          }
+          else
+          {
+            memcpy(charTmpBuf, dataPtr, tokenLen);
+          }
+        }
+
+        uint64_t compChar = uint64ToStr(*(reinterpret_cast<uint64_t*>(charTmpBuf)));
+        int64_t binChar = static_cast<int64_t>(compChar);
+
+        // Update min/max range
+        uint64_t minVal = static_cast<uint64_t>(bufStats.minBufferVal);
+        uint64_t maxVal = static_cast<uint64_t>(bufStats.maxBufferVal);
+
+        if (compChar < minVal)
+          bufStats.minBufferVal = binChar;
+        if (compChar > maxVal)
+          bufStats.maxBufferVal = binChar;
+
+        pVal = charTmpBuf;
+        memcpy(p, pVal, width);
+      }
+      break;
+    }
+
+    //----------------------------------------------------------------------
+    // SHORT INT
+    //----------------------------------------------------------------------
+    case WriteEngine::WR_SHORT:
+    {
+      long long origVal;
+      // use char type here
+      const short* dataPtr = columnData->data()->GetValues<short>(1);
+      for (unsigned int i = 0; i < cbs; i++)
+      {
+        bool bSatVal = false;
+        void* p = buf + i * width;
+        if (columnData->IsNull(i))
+        {
+          if (!column.autoIncFlag)
+          {
+            if (column.fWithDefault)
+            {
+              origVal = column.fDefaultInt;
+            }
+            else
+            {
+              siVal = joblist::SMALLINTNULL;
+              pVal = &siVal;
+              memcpy(p, pVal, width);
+              continue;
+            }
+          }
+          else
+          {
+            // fill 1 temporarily
+            origVal = fAutoIncNextValue++;
+          }
+        }
+        else
+        {
+          if ((column.dataType == CalpontSystemCatalog::DECIMAL) ||
+              (column.dataType == CalpontSystemCatalog::UDECIMAL))
+          {
+            const int128_t* dataPtr1 = reinterpret_cast<const int128_t*>(dataPtr);
+            origVal = *(dataPtr1 + i);
```

Comment:
Indeed, this `if-else` block cannot be omitted, when the imported data is in the length of `short int`, it will store in `short int` type. So the data should first be cast to `int128` then store in `origVal`.

Reply:
No, this is wrong. ColumnStore internally has 5 different types for DECIMAL, depending on the DECIMAL precision:
```
  tiny int (1 byte),
  short int (2 bytes),
  int (4 bytes),
  big int (8 bytes),
  int128_t (16 bytes)
```
This case block is for SHORT INT, so there is no need to first cast to `int128_t*`. Moreover, this type casting would segfault because the input `dataPtr` has a `short` type.

Reply:
I understand ColumnStore store different types DECIMAL depending on the DECIMAL precision. But for decimal128 type in parquet, each data takes 128bit to store. So here though initially `dataPtr` has a `short` type, it's not true. It later should be cast as `int128_t` cause every element in array takes 128 bits. If there is no casting operation, `origVal = *(dataPtr + i)` will get 16 bits one by one which is not correct.

Reply:
> No, this is wrong. ColumnStore internally has 5 different types for DECIMAL, depending on the DECIMAL precision:
> 
> ```
>   tiny int (1 byte),
>   short int (2 bytes),
>   int (4 bytes),
>   big int (8 bytes),
>   int128_t (16 bytes)
> ```
> 
> This case block is for SHORT INT, so there is no need to first cast to `int128_t*`. Moreover, this type casting would segfault because the input `dataPtr` has a `short` type.
> 
> Please also fix this for other cases below where I have added a comment.

I understand ColumnStore store different types DECIMAL depending on the DECIMAL precision. But for decimal128 type in parquet, each data takes 128bit to store. So here though initially `dataPtr` has a `short` type, it's not true. It later should be cast as `int128_t` cause every element in array takes 128 bits. If there is no casting operation, `origVal = *(dataPtr + i)` will get 16 bits one by one which is not correct.

Reply:
Are you sure this case block is for `decimal128` type in parquet? This case block is `case WriteEngine::WR_SHORT`. So I am assuming the corresponding parquet type will be a 2-byte (16-bit) type here and not 128-bits. I don't believe all parquet DECIMAL types are stored using 128 bits, smaller DECIMAL types should be using less bytes for storage. Can you please double check the parquet doc/spec on this discussion and post a link to the spec here?

Your case block handling in `case WriteEngine::WR_BINARY` correctly handles the 128-bit DECIMAL type using the `arrow::Decimal128Array` type.

Reply:
> Are you sure this case block is for `decimal128` type in parquet? This case block is `case WriteEngine::WR_SHORT`. So I am assuming the corresponding parquet type will be a 2-byte (16-bit) type here and not 128-bits. I don't believe all parquet DECIMAL types are stored using 128 bits, smaller DECIMAL types should be using less bytes for storage. Can you please double check the parquet doc/spec on this discussion and post a link to the spec here?
> 
> Your case block handling in `case WriteEngine::WR_BINARY` correctly handles the 128-bit DECIMAL type using the `arrow::Decimal128Array` type.

In parquet, it only provides decimal128 and decimal256. And in columnstore here, for `WR_BINARY`, it's data with 128 bits thus I use decimal128 to create decimal column data. Except these two decimal type, parquet doesn't provide decimal consistent with `short` or other types. So if a column in parquet is decimal type, I defaultly think it's decimal 128 or its data length cannot be judged correctly.
[https://arrow.apache.org/docs/12.0/cpp/api/datatype.html#_CPPv4N5arrow14Decimal128TypeE](https://arrow.apache.org/docs/12.0/cpp/api/datatype.html#_CPPv4N5arrow14Decimal128TypeE)